### ISSUE TITLE
[MNG-1867] deprecate system scope, analyse other use cases

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#3356]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** opened **[MNG-1867](https://issues.apache.org/jira/browse/MNG-1867?redirect=false)** and commented

possibly can avoid all use cases for system scope through proper use of alternate resolvers. Gather use cases (see MNG-1471) to ensure.

---

**Attachments:**
- [MNG-1867.zip](https://issues.apache.org/jira/secure/attachment/12712738/MNG-1867.zip) (_11.30 kB_)

**Issue Links:**
- [MNG-2070](https://issues.apache.org/jira/browse/MNG-2070) systemPath is needed for tests as well as compile time scope
 (_**"is duplicated by"**_)
- [MCOMPILER-3](https://issues.apache.org/jira/browse/MCOMPILER-3) Module paths for system scope are relative to parent pom instead of its own
 (_**"depends upon"**_)

**Remote Links:**
- [Page
](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=65875544)

3 votes, 15 watchers

**Comment by jira-importer** [comment]
**[Mike Whittemore](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mikewhit)** commented

I would like to be able to specify a system path for any dependency regardless of the scope. For example, I would like to mark junit-3.8.1 as being in the "test" scope, yet point at a jar that is not in the repository. Something like this:

\<dependency>
\<groupId>junit\</groupId>
\<artifactId>junit\</artifactId>
\<version>3.8.1\</version>
\<scope>test\</scope>
\<path>${cots.top}/junit/junit.jar\</path>
\</dependency>

I don't want to mark it as "system" scope, simply to say it is not in the repository because that would lose the fact that it is really something scoped for the unit test phase.

I understand the preferred way to use maven is to install 3rd-party artifacts in the repository (and the public ibiblio repository if possible), but the security culture in my company will not allow it (there is a rigid procedure for how 3rd-party jars and tools get downloaded, installed, and used -  not uncommon in our particular industry). Without this feature I will be forced to mark all 3rd-party jars as "system" scoped, which I want to avoid. Regardless of my special circumstances, this feature would add a little more flexibility to maven for handling special cases.

So to summarize, I would like to be able to tag certain dependencies with a system path, but still tag them with an appropriate build scope such as"runtime",  "test" or "provided". If tagged with a path, the dependency would not go to the repository to resolve the jar, but instead would follow the system path. Additionally, the path should be allowed to contain variables so I can make the build portable across operating systems (for example, by using a varaible to indicate the root directory of COTS software installations).

I would be happy to help out with the implementation, though I would need a few pointers of where to focus my attention.

**Comment by jira-importer** [comment]
**[Brett Porter](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=brettporter)** commented

Mike - we are currently resorting to custom resolvers for such cases. You basically use the repository, or you don't - it can't be mixed and matched. Is this acceptable if it is documented?

**Comment by jira-importer** [comment]
**[Mike Whittemore](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mikewhit)** commented

Sorry, I'm not sure what you mean by custom resolver. I browsed the subversion files a bit and nothing really jumped out at me. Are you saying there is something called a "custom resolver" that I can configure such that Maven will resolve dependecies found outside the repository, and that if I use this mechanism I cannot use the repository at all? If so, where do build artifacts go when I run Maven install?

If you are saying there is essentially a way to do what I need but it requires giving up use of the repository completely, then documenting that would likely solve my dilemma. Thanks.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

Deprecate during 3.0, remove in 3.1. System scope and picking artifacts up off the local file system is bad.

**Comment by jira-importer** [comment]
**[Jesse N. Glick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jglick)** commented

System scope is the only known way to do certain tasks. Most notably:

http://maven.apache.org/general.html#tools-jar-dependency

This case could be solved "properly" only if everyone agreed that the (Open)JDK 6 or 7 version of `tools.jar` sufficed, and could be legally published to a well-known repository. There may be other rarer cases with `dt.jar` etc.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

It's not the only way. What we do in the compiler plugin is discovery the location and load it appropriately. One small library to deal with the locations that JDK provides on various platforms. We're not putting a JVM specific tools.jar in Maven Central. This doesn't work for people who use the IBM JDK, or deal with cases where platforms have custom JVMs like the Blackberry. The strategy employed in the compiler plugin will be expanded. A little more work, but the right thing to do.

**Comment by jira-importer** [comment]
**[Jesse N. Glick](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jglick)** commented

Most of `tools.jar` - the part created by the OpenJDK langtools project - is not JVM-specific and can be used as a plain old Java library with any JRE (1.5 or later). For example, if you want an implementation of `javax.tools.JavaCompiler` to be present with test scope (because your unit tests generate and compile source code on the fly), this would be the obvious choice. Besides clarity, using regular artifacts provides better reproducibility of builds since the version of the tool is encoded directly in the POM in the normal way rather than being at the mercy of the environment, and you could switch to a particular version of ECJ if that were faster or did not have a crucial bug.

There are some less frequently used pieces like the `com.sun.tools.attach` and `com.sun.jdi` APIs which should probably be packaged as separate artifacts with the understanding that they are applicable to JREs derived from the Sun codebase. You ought to be able to compile modules using these APIs without requiring Maven to be run with a Sun JDK.

Special embedded JVMs generally need their own tools. For example, for Android development the plugin handles the special build, packaging, and verification steps needed by that platform. This would not be done with system-scope JARs to begin with, and the JVM running Maven is almost certainly different from the target JVM.

**Comment by jira-importer** [comment]
**[Karl Heinz Marbaise](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=khmarbaise)** commented

Currently i have the following example of using the system scope:

```
<dependency>
  <groupId>org.apache.hadoop</groupId>
  <artifactId>hadoop-common</artifactId>
  <version>${hadoop.version}</version>
  <scope>system</scope>
  <systemPath>${basedir}/hadoop-${hadoop.version}/hadoop-common-${hadoop.version}.jar</systemPath>
</dependency>
...
```

This makes it possible to handle the situation where artifacts are not availabe as Maven artifacts (central..). I would suggest to make at least a scope available to have dependencies which are located within the current project.

```
<dependency>
  <...>
  <scope>project</scope>
  <systemPath>../...jar</systemPath>
</dependency>
```

Furthermore the usage of things like tools.jar of the JDK would be usefull as well as mentioned earlier.

**Comment by jira-importer** [comment]
**[Jason van Zyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl)** commented

I think Tycho is the last legitimate use case for scoped dependencies. If that were taken care I think we could wipe out system scope.

**Comment by jira-importer** [comment]
**[Kevin Neal](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=freebeer)** commented

Silk4J is another use case. Silk installs a jar onto a local path (C:\Program Files (x86)\Silk\SilkTest\ng\JTF\silktest-jtf-nodeps.jar... yes, it's Windows only) that is specifically able to communicate with an agent installed on the local machine.

The jar can be installed in the local repo or deployed to a private repository, but I haven't figured out another way to get a version of the jar that is compatible with the local agent. I understand there are drawbacks for Maven supporting local system paths, but it's a necessary evil for this situation.

UPDATE - In case this is the last remaining use case, feel free to ignore since the company this was at decided to close.

**Comment by jira-importer** [comment]
**[Juan Farré](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jafarre.insags)** commented

My only two use cases for system scope are the well-known tools.jar and getting rid of transitive dependencies which are now incluided in the JRE.
Many maven artifacts carry transitive dependencies en compile scope with things like XML API, JAXB, and many other JAR's which are included in JRE 7.
I usually manage these dependencies setting system scope to rt.jar so that they are not included when packaging (this could be achieved with scope provided), nor when compiling or testing.
I don't see the alternative, unless I take the huge work of adding the exclusions.
Is there an alternative for this use case? What do you suggest?
By the way, all these rt.jar dependencies fail with JRE 9 and I do not even know how to fix them. Luckily, not using Java 9 in any project, yet...
